src/main/java/com/wire/bots/hold/service/DeviceManagementService.java

```diff
@@ -55,17 +70,75 @@ public InitializedDeviceDTO initiateLegalHoldDevice(QualifiedId userId, UUID tea
      * </p>
      * @param userId user setup to be put under legal hold
      * @param teamId user's own team
      * @param clientId user's device
      * @param refreshToken token used to get expiring api tokens
+     * @throws RuntimeException when any of the (parallel or not) MLS tasks fails. Or if inserting refreshToken to Database fails.
      */
-    public void confirmDevice(QualifiedId userId, UUID teamId, String clientId, String refreshToken) {
+    public void confirmDevice(QualifiedId userId, UUID teamId, String clientId, String refreshToken) throws RuntimeException {
+        API api = new API(client, null, refreshToken);
+        if (api.isMlsEnabled()) {
+            try (CryptoMlsClient cryptoMlsClient = new CryptoMlsClient(clientId, coreCryptoPassword)) {
+                // CryptoMlsClient will be closed from `try` with resource so there is no issue passing
+                // Crypto as null, as we will not be calling wireClientBase.close()
+                WireClientBase wireClientBase = new WireClientBase(api, null, cryptoMlsClient, null);
+
+                CompletableFuture<Void> mlsPublicKeyFuture = CompletableFuture.supplyAsync(() -> {
+                    wireClientBase.updateClientWithMlsPublicKey();
+                    return null;
+                });
+                CompletableFuture<Void> mlsKeyPackagesFuture = CompletableFuture.supplyAsync(() -> {
+                    wireClientBase.uploadMlsKeyPackages(100);
+                    return null;
+                });
+                CompletableFuture<List<Conversation>> conversationsFuture = CompletableFuture.supplyAsync(() ->
+                    api.getUserConversations()
+                   .stream()
+                   .filter(c -> c.protocol == Conversation.Protocol.MLS)
+                   .collect(Collectors.toList()));
+
+                CompletableFuture<Void> combinedFutures = CompletableFuture.allOf(
+                    mlsPublicKeyFuture,
+                    mlsKeyPackagesFuture,
+                    conversationsFuture
+                ).handle((success, throwable) -> {
+                    if (throwable != null) {
+                        if (throwable instanceof CompletionException) {
+                            throw new RuntimeException(throwable.getCause().getMessage());
+                        } else {
+                            throw new RuntimeException(throwable.getMessage());
+                        }
+                    }
+
+                    return success;
+                });
+
+                combinedFutures.get();
+
+                for (Conversation conversation : conversationsFuture.get()) {
+                    wireClientBase.joinMlsConversation(conversation.id, conversation.mlsGroupId);
+                }
+            } catch (ExecutionException exception) {
+                throw new RuntimeException(exception.getCause().getMessage());
+            } catch (Exception exception) {
```

Comment:
Not sure the second catch is needed, catching any exception and re-throwing a new one makes sense if on a upper layer you have something that catches specifically a RuntimeException. With this approach, on the ConfirmResourceV1 and V0 you the logger with put the correct message, but you lose the information about which kind of exception it was.
If you just remove the second catch, the exception will propagate in the same way but you also keep the exception type

Reply:
This was changed from `InterruptedException` to a generic `Exception` because:

`combinedFutures.get();` or `conversationsFuture.get();` throws a `InterruptedException`, but any exception from `wireClientBase.joinMlsConversation(..)` will throw a `RuntimeException` but having all explicit would be just copying and pasting the same code, like this:

```
 catch (ExecutionException exception) {
      throw new RuntimeException(exception.getCause().getMessage());
} catch (InterruptedException exception) {
      throw new RuntimeException(exception.getMessage());
} catch (Exception exception) {
      throw new RuntimeException(exception.getMessage());
}
```

I'm fine with removing the generic `Execption` but then we would need to add back the `InterruptedException` catch

Reply:
Aaaaah I understand, then it's quite similar.
I would still go with replacing the second catch from Exception to InterruptedException then.
And maybe in both catches we can do something like `throw new RuntimeException("ExecutionException" + exception.getCause().getMessage());`
Most likely the message will contain refererences to the exception type but we can never be sure